src/main/scala/org/broadinstitute/dsde/workbench/sam/Boot.scala

```diff
@@ -339,12 +340,56 @@ object Boot extends IOApp with LazyLogging {
       accessPolicyDAO: AccessPolicyDAO,
       directoryDAO: DirectoryDAO)(implicit actorSystem: ActorSystem, actorMaterializer: ActorMaterializer): AppDependencies = {
     val resourceTypeMap = config.resourceTypes.map(rt => rt.name -> rt).toMap
-    val policyEvaluatorService = PolicyEvaluatorService(config.emailDomain, resourceTypeMap, accessPolicyDAO, directoryDAO)
-    val resourceService = new ResourceService(resourceTypeMap, policyEvaluatorService, accessPolicyDAO, directoryDAO, cloudExtensionsInitializer.cloudExtensions, config.emailDomain)
-    val userService = new UserService(directoryDAO, cloudExtensionsInitializer.cloudExtensions)
-    val statusService = new StatusService(directoryDAO, cloudExtensionsInitializer.cloudExtensions, 10 seconds)
-    val managedGroupService =
-      new ManagedGroupService(resourceService, policyEvaluatorService, resourceTypeMap, accessPolicyDAO, directoryDAO, cloudExtensionsInitializer.cloudExtensions, config.emailDomain)
+
+    val (policyEvaluatorService, resourceService, userService, statusService, managedGroupService) =
+      if (Proxy.isProxyClass(accessPolicyDAO.getClass)) {
+        val accessPolicyHandler = Proxy.getInvocationHandler(accessPolicyDAO).asInstanceOf[ShadowRunnerDynamicProxy[AccessPolicyDAO]]
+        val directoryHandler = Proxy.getInvocationHandler(directoryDAO).asInstanceOf[ShadowRunnerDynamicProxy[DirectoryDAO]]
+
+        val realPolicyEvalSvc = PolicyEvaluatorService(config.emailDomain, resourceTypeMap, accessPolicyHandler.realDAO, directoryHandler.realDAO)
+        val shadowPolicyEvalSvc = PolicyEvaluatorService(config.emailDomain, resourceTypeMap, accessPolicyHandler.shadowDAO, directoryHandler.shadowDAO)
+        val realResourceService = ResourceService(resourceTypeMap, realPolicyEvalSvc, accessPolicyHandler.realDAO, directoryHandler.realDAO, cloudExtensionsInitializer.cloudExtensions, config.emailDomain)
+        val shadowResourceService = ResourceService(resourceTypeMap, shadowPolicyEvalSvc, accessPolicyHandler.shadowDAO, directoryHandler.shadowDAO, NoExtensions, config.emailDomain)
+
+        val policyEvaluatorService = DaoWithShadow[PolicyEvaluatorService](
+          realPolicyEvalSvc,
+          shadowPolicyEvalSvc,
+          accessPolicyHandler.resultReporter, accessPolicyHandler.clock
```

Comment:
Is there any value in creating a new `resultReporter` here and in the other services with a different `daoName` to have more accurate metrics or is that more trouble than it's worth?